core/src/main/scala/kafka/server/ReplicaAlterLogDirsThread.scala

```diff
@@ -76,13 +82,49 @@ class ReplicaAlterLogDirsThread(name: String,
     futureLog.updateHighWatermark(partitionData.highWatermark)
     futureLog.maybeIncrementLogStartOffset(partitionData.logStartOffset, LogStartOffsetIncrementReason.LeaderOffsetIncremented)
 
-    if (partition.maybeReplaceCurrentWithFutureReplica())
-      removePartitions(Set(topicPartition))
+    directoryEventHandler match {
+      case DirectoryEventHandler.NOOP =>
+        if (partition.maybeReplaceCurrentWithFutureReplica())
+          removePartitions(Set(topicPartition))
+      case _ =>
+        maybePromoteFutureReplica(topicPartition, partition)
+    }
```

Comment:
Perhaps we make DirectoryEventHandler.NOOP immediately callback with success so we don't need this extra logic.

Reply:
Which callback you are referring to here? The order of checking partition future replica is different between DirectoryEventHandler.NOOP and when we have a handler.

Reply:
I was suggesting that NOOP immediatly calls the `Runnable callback` in `handleAssignment()`, but given the current logic in `maybePromoteFutureReplica` think my suggestion isn't as straightforward as I was imagining.